netlogo-gui/src/main/ide/JumpToDeclaration.scala

```diff
@@ -0,0 +1,91 @@
+    iterator.find(p => p.start < position && p.end >= position)
+  }
+  def jumpToDeclaration(cursorPosition: Int, editorArea: EditorArea): Unit = {
+    val tokenOption = findTokenContainingPosition(editorArea.getText(), cursorPosition)
+    for(token <- tokenOption) {
+      val t = getDeclaration(token, editorArea.getText)
+      t.foreach(token => {
+        editorArea.select(token.start, token.end)
+      })
+    }
+  }
+
+  def getDeclaration(token: Token, source: String): Option[Token] = {
+    val tokens = Femto.scalaSingleton[TokenizerInterface]("org.nlogo.lex.Tokenizer").tokenizeString(source).toSeq
+    val (globals, owns) = parseOwns(tokens)
+    val indexOfToken = tokens.indexWhere(t => t.start == token.start)
+    var stack = 0
+    for(i <- indexOfToken to 0 by -1) {
+      val tok = tokens(i)
+      if (tok.tpe.equals(TokenType.CloseBracket)) {
+        stack -= 1
+      }
+      if (tok.tpe.equals(TokenType.OpenBracket)) {
+        stack += 1
+      }
+      if(stack == 0 && tok.text.equalsIgnoreCase(token.text) && tokens(i - 1).text.equalsIgnoreCase("let")) {
+        return Some(tok)
+      }
+      if(tok.text.equalsIgnoreCase("to") || tok.text.equalsIgnoreCase("to-report")) {
+        // i + 1 is the function name
+        if(tokens(i + 2).tpe.equals(TokenType.OpenBracket)){
+          // Arguments found
+          var argIndex = i + 3
+          var arg = tokens(argIndex)
+          while(!tokens(argIndex).tpe.equals(TokenType.CloseBracket)){
+            if(tokens(argIndex).text.equalsIgnoreCase(token.text)) {
+              return Some(tokens(argIndex))
+            }
+            argIndex += 1
+          }
+        }
+        globals.get(token.text.toUpperCase) orElse owns.get(token.text.toUpperCase)
+      }
+    }
+    globals.get(token.text.toUpperCase) orElse owns.get(token.text.toUpperCase)
+  }
+
+  def parseOwns(tokens: Seq[Token]): (scala.collection.mutable.Map[String, Token], scala.collection.mutable.Map[String, Token]) = {
+    val globals = scala.collection.mutable.Map[String, Token]()
+    val owns = scala.collection.mutable.Map[String, Token]()
+    var i = 0
+    while(i < tokens.length) {
+      var token = tokens(i)
+      if (token.text.equalsIgnoreCase("GLOBALS")) {
+        i += 1
+        token = tokens(i)
+        while (!token.tpe.equals(TokenType.CloseBracket) && i < tokens.length) {
+          if (token.tpe.equals(TokenType.Ident))
+            globals += token.text.toUpperCase -> token
+          token = tokens(i)
+          i += 1
+        }
+      }
+      if (token.text.toUpperCase.endsWith("-OWN")) {
+        i += 1
+        token = tokens(i)
+        while (!token.tpe.equals(TokenType.CloseBracket) && i < tokens.length) {
+          if (token.tpe.equals(TokenType.Ident) && owns.get(token.text).isEmpty) {
+            owns += token.text.toUpperCase -> token
+          }
+          token = tokens(i)
+          i += 1
+        }
+      }
+      i += 1
+    }
```

Comment:
I see you have stopped using `iterator` here, which is good. There are still cases where this can throw an exception. For instance, if I have a code tab with just `globals` (or `globals [` or `globals [ foo `) and I right-click and choose "jump to definition", I see an `OutOfBoundsException`.

Anytime you're indexing into a sequence, make sure the sequence is defined at that point. Part of the reason that list-comprehensions are nice is that they check for element existence for you. What you're doing here could be rewritten as a `foldLeft`, although it would be quite complex. In this case it might be enough to just check before each index into the list that index is less than the length of the list.